[kuiktzecheng] iP

kuiktzecheng opened the pull request:
### **Gary - Task Manager Chatbot**

> "The secret of getting ahead is getting started." – Mark Twain  [[source]](https://www.goodreads.com/quotes/219455-the-secret-of-getting-ahead-is-getting-started-the-secret)

Gary is here to help you manage your tasks effortlessly in a simple, chat-based interface. It’s:

- Fast and intuitive
- Focused on task management
- Always ready to help you stay organized 📝

All you need to do is:

1. Download Gary from [here](https://github.com/kuiktzecheng/ip/releases/tag/A-JAr).
2. Start chatting with Gary.
3. Add tasks and organize them.
4. Let Gary manage your to-do list for you!
5. And yes, it’s **_FREE!_**

Features:

- [X]  Managing task lists
- [X]  Setting task deadlines
- [ ]  Adding reminders (coming soon)


If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

<pre> public class Gary {
    public static void main(String[] args) {
        new Gary("src/textFile/gary.txt").start();
    }
}
</pre>

Commit:
Add Gradle support

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

kuiktzecheng committed:
Add Gary.java

kuiktzecheng committed:
Add Gary.java

kuiktzecheng committed:
Added Echo Feature

kuiktzecheng committed:
Added Echo Feature

kuiktzecheng committed:
Add, List Feature

kuiktzecheng committed:
Add, List Feature

kuiktzecheng committed:
Mark as Done Feature

kuiktzecheng committed:
Mark as Done Feature

kuiktzecheng committed:
Todo, Event, Dealine Feature

kuiktzecheng committed:
Todo, Event, Dealine Feature

kuiktzecheng committed:
Todo, Deadline, Event feature

kuiktzecheng committed:
Todo, Deadline, Event feature

kuiktzecheng committed:
Automated Text UI Testing

kuiktzecheng committed:
Automated Text UI Testing

kuiktzecheng committed:
Automated Text UI Testing

kuiktzecheng committed:
Automated Text UI Testing

kuiktzecheng committed:
Change test file

kuiktzecheng committed:
Change test file

kuiktzecheng committed:
Handled Exceptions

kuiktzecheng committed:
Handled Exceptions

kuiktzecheng committed:
Added Delete Feature

kuiktzecheng committed:
Added Delete Feature

kuiktzecheng committed:
Add save feature

kuiktzecheng committed:
Add save feature

kuiktzecheng committed:
Merge branch-Level-7 to master

kuiktzecheng committed:
Add dates and times feature

kuiktzecheng committed:
Add dates and times feature

kuiktzecheng committed:
Merge branch-Level-8 to master

kuiktzecheng committed:
Use More OOP

kuiktzecheng committed:
Use More OOP

kuiktzecheng committed:
Merge branch-AMoreOOP to master

kuiktzecheng committed:
Divide classes into packages

kuiktzecheng committed:
Divide classes into packages

kuiktzecheng committed:
Merge branch-A-Packages to master

kuiktzecheng committed:
Remove redundant files

kuiktzecheng committed:
Merge branch 'add-gradle-support'

kuiktzecheng committed:
Add JUnit Tests

kuiktzecheng committed:
Merge branch-A-JUnit to master

yu7ong commented on src/main/java/Gary/Ui.java:
```diff
@@ -0,0 +1,46 @@
+    public Ui() {
+        this.detector = new Scanner(System.in);
+    }
+    public void greet() {
+        System.out.println("Hello! I'm Gary\nWhat can I do for you?");
+    }
+    public void goodbye() {
+        this.detector.close();
+        System.out.println("Bye. Hope to see you again soon!");
+    }
+    public void markTask(Task task) {
+        System.out.println("Nice! I've marked this task as done:");
+        System.out.println("\t" + task.toString());
+    }
+    public void unmarkTask(Task task) {
+        System.out.println("OK, I've marked this task as not done yet:");
+        System.out.println("\t" + task.toString());
+    }
+    public void deleteTask(Task task, int size) {
+        System.out.println("Noted. I've removed this task from the list:\n " + task.toString() +
+                "\nNow you have " + size + " tasks in the list.\n");
+    }
+    public void addTask(Task task, int size) {
+        System.out.println("Got it. I've added this task:");
+        System.out.println("\t" + task.toString());
+        System.out.println("Now you have " + size + " tasks in the list.");
+    }
+    public void showTaskLists(TaskList taskList) {
+        for (int i = 0; i < taskList.size(); i++) {
+            System.out.println(i + 1 + "." + taskList.getTask(i));
+        }
+    }
+    public String read() {
+        return detector.nextLine();
+    }
+    public void showError(String message) {
+        System.out.println(message);
+    }
```
Naming of methods is good.

yu7ong commented on src/main/java/Gary/TaskList.java:
```diff
@@ -0,0 +1,67 @@
+    private ArrayList<Task> taskLists;
+
+    public TaskList() {
+        this.taskLists = new ArrayList<>();
```
Perhaps 'tasks' or 'taskList' will be clearer here?

yu7ong commented on src/main/java/Gary/task/Event.java:
```diff
@@ -0,0 +1,49 @@
+import java.time.format.DateTimeFormatter;
+
+public class Event extends Task {
+    protected LocalDateTime start;
+    protected LocalDateTime end;
```
I like how distinguishing names are used.

yu7ong commented on src/main/java/Gary/Parser.java:
```diff
@@ -0,0 +1,76 @@
+                throw new GaryException("Please provide your Deadline task in the following format: \n" +
+                        "Gary.task.Deadline <task name> /by <yyyy-mm-dd>\n");
+            }
+            String description = split1[0].trim();
+            String dueDate = split1[1].trim();
```
I like how the names explain the entity to the readers.

yu7ong reviewed (commented):
LGTM

kuiktzecheng committed:
Add JavaDocs for classes and methods

kuiktzecheng committed:
Follow Coding Standard

blanklogic commented on src/main/java/Gary/Gary.java:
```diff
@@ -0,0 +1,38 @@
+
+import java.io.FileNotFoundException;
+import Gary.command.Command;
+public class Gary {
```
Consider adding a whitespace between imports and the start of the class as well as including JavaDocs of what the class as well as it's methods aim to do for better documentation of your code.

blanklogic commented on src/main/java/Gary/Gary.java:
```diff
@@ -0,0 +1,38 @@
+package Gary;
+
+import java.io.FileNotFoundException;
+import Gary.command.Command;
```
Checkstyle issue: requires whitespace between java import and Gary import

blanklogic commented on src/main/java/Gary/Parser.java:
```diff
@@ -0,0 +1,76 @@
+package Gary;
+
+import Gary.command.AddCommand;
```
Perhaps having your Parser.java in Gary/command might remove the need for the imports from Lines 3 to 8.

blanklogic commented on src/main/java/Gary/Parser.java:
```diff
@@ -0,0 +1,76 @@
+import Gary.task.ToDo;
+import Gary.task.Deadline;
+import Gary.task.Event;
+public class Parser {
```
Java Coding Standard Violation: Whitespace between imports and class

blanklogic commented on src/main/java/Gary/Parser.java:
```diff
@@ -0,0 +1,76 @@
+import Gary.task.Deadline;
+import Gary.task.Event;
+public class Parser {
+    static Command parse(String fullCommand) throws GaryException {
```
Access type not specified. Suggested access type to be public.

blanklogic commented on src/main/java/Gary/Parser.java:
```diff
@@ -0,0 +1,76 @@
+            }
+            return new AddCommand(new Deadline(description, dueDate));
+        case ("event") :
+            if (split.length != 2) {
```
Consider looking up and utilising regex to reduce the need for first split and second split, reducing the amount of repeated code as a good Software Engineering Practice: Don't Repeat Yourself

blanklogic commented on src/main/java/Gary/Parser.java:
```diff
@@ -0,0 +1,76 @@
+        case "bye":
+            return new ByeCommand();
+        }
+         throw new GaryException("Sorry! I do not understand what is this!!");
```
Wrong Indentation Level.

blanklogic commented on src/main/java/Gary/GaryException.java:
```diff
@@ -0,0 +1,7 @@
+package Gary;
+
+public class GaryException extends Exception {
```
Perhaps extending from RuntimeException is more appropriate as GaryException is only thrown during Runtime as opposed to Compile Time (Exception).

blanklogic commented on src/main/java/Gary/Storage.java:
```diff
@@ -0,0 +1,44 @@
+import java.io.IOException;
+import java.util.Scanner;
+import Gary.task.Task;
+public class Storage {
```
Java Coding Standard Violation: Whitespace between imports and class

blanklogic commented on src/main/java/Gary/TaskList.java:
```diff
@@ -0,0 +1,67 @@
+import Gary.task.Deadline;
+import Gary.task.Event;
+
+
```
Java Coding Standard Violation: Excessive Whitespace between imports and class

blanklogic commented on src/main/java/Gary/TaskList.java:
```diff
@@ -0,0 +1,67 @@
+
+
+public class TaskList {
+    private ArrayList<Task> taskLists;
```
Can consider making taskLists effectively final since after being initialised by the constructor, it's reference is not modified. Also naming suggestion: tasks might be easier to understand as a list of tasks instead of multiple task lists.

blanklogic commented on src/main/java/Gary/Ui.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.Scanner;
+import Gary.task.Task;
+public class Ui {
```
Java Coding Standard Violation: Whitespace between imports and class

blanklogic commented on src/main/java/Gary/command/AddCommand.java:
```diff
@@ -0,0 +1,36 @@
+            ui.addTask(task, taskLists.size());
+            try {
+                storage.saveTask(taskLists);
+            } catch(IOException e) {
```
Catch block is empty with no appropriate resolution provided.

blanklogic commented on src/main/java/Gary/command/DeleteCommand.java:
```diff
@@ -0,0 +1,33 @@
+import Gary.task.Task;
+
+public class DeleteCommand extends Command {
+    int deletedIndex;
```
Access type not provided which might cause leakage of information hiding. Suggestion: private

blanklogic commented on src/main/java/Gary/command/EditTaskCommand.java:
```diff
@@ -0,0 +1,40 @@
+        }
+
+    }
+    @Override
```
I noticed a consistent repeat of this Overriding of isBye(). Consider making isBye() in abstract class Command.java to return false and only Override the ByeCommand to return true. Software Engineering Practice: Don't Repeat Yourself.

blanklogic commented on src/main/java/Gary/task/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+public class Deadline extends Task {
```
Java Coding Standard Violation: Whitespace between imports and class

blanklogic commented on src/main/java/Gary/task/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+        return LocalDate.parse(dateTime, inputFormatter);
+    }
+    @Override
+    public String toString() {
```
Task.java has already accomplished the following, consider calling super.toString() for this part to reduce the need to repeat yourself in your code.
```java
"[" + (this.isDone ? "X" : " ") + "] " + this.description;
```

blanklogic commented on src/main/java/Gary/task/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+    private LocalDate deadline;
+
+    static DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern("MMM dd yyyy");
+    static DateTimeFormatter inputFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
```
Access type not specified. Suggestion: private

blanklogic commented on src/main/java/Gary/task/Event.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Override
+    public String toString() {
+        return "[E][" + (this.isDone ? "X" : " ") + "] " + this.description +
```
Task.java has already accomplished the following, consider calling super.toString() for this part to reduce the need to repeat yourself in your code.
```java
"[" + (this.isDone ? "X" : " ") + "] " + this.description;
```

blanklogic commented on src/textFile/gary.txt:
```diff
@@ -0,0 +1,3 @@
+T | 0 | buy bread
```
gary.txt should be added to gitignore as it is constantly changing and should not be tracked. Do consider removing this cached version as well to prevent further tracking.

blanklogic reviewed (commented):
# Overview
There seems to be a lot of Java Coding Standard Violations which I suggest you read up and make the appropriate fixes to the suggestions provided. Furthermore, the lack of JavaDoc makes it difficult for reviewers to easily understand what each class and its methods are doing. There is also quite a significant amount of overlapping codes which could be optimised through abstraction and inheritance. Do consider reading up on OOP Principles (Inheritance, Abstraction, Information Hiding, Tell Don't Ask), good Software Engineering Practices (DRY, SOLID) and drawing a UML diagram to better visualise.

# Conclusion
Changes are required to improve code quality and code documentation to maintain existing code base integrity. Another review might be required prior to merging.

kuiktzecheng committed:
Add Final Changes

kuiktzecheng committed:
Add find feature

kuiktzecheng committed:
Resolve merge conflicts for branch-A-JavaDoc

kuiktzecheng committed:
Resolve merge conflicts for branch-A-CodingStandard

kuiktzecheng committed:
Resolve merge conflicts for branch-Level-9

kuiktzecheng committed:
Style code according to coding standard

kuiktzecheng committed:
Merge branch-A-Checkstyle to master

kuiktzecheng committed:
Add GUI to chatbot

kuiktzecheng committed:
Merge branch-Level-10 to master

kuiktzecheng committed:
Add assertions to Gary components

The current code lacks assertions, which makes it vulnerable to
hidden errors and unexpected behaviors during runtime. This may
increase the difficulty of debugging when invalid states occur,
leading to larger issues down the line.

Introduce assertions in the various files to validate internal
states and inputs. Use assertions to check conditions before
proceeding with operations, enhancing code reliability and
maintainability.

Assertions are implemented because they provide a straightforward
and effective method for catching errors early without significantly
altering the existing code structure. This helps to identify potential
issues during the development phase, thereby aiding in maintaining
overall code quality.

kuiktzecheng committed:
Merge pull request #2 from kuiktzecheng/branch-A-Assertions

Add assertions to Gary components

kuiktzecheng committed:
Add changes

kuiktzecheng committed:
Refactor codebase to improve code quality

Critically examined the codebase and refactored various classes and methods
to enhance code quality. This includes:

- Improving field immutability by marking variables as final where appropriate.
- Enhancing error handling by providing more context in exception messages.
- Removing redundant code and simplifying logic in methods.
- Adding and improving Javadoc comments for better documentation and clarity.

These changes aim to make the code more maintainable, readable, and robust.

kuiktzecheng committed:
Refactor codebase to improve code quality

Critically examined the codebase and refactored various classes and methods
to enhance code quality. This includes:

- Improving field immutability by marking variables as final where appropriate.
- Enhancing error handling by providing more context in exception messages.
- Removing redundant code and simplifying logic in methods.
- Adding and improving Javadoc comments for better documentation and clarity.

These changes aim to make the code more maintainable, readable, and robust.

kuiktzecheng committed:
Merge pull request #3 from kuiktzecheng/branch-A-CodeQuality

Branch A-CodeQuality

kuiktzecheng committed:
Add detect duplicate extension to chatbot

kuiktzecheng committed:
Merge branch-C-DetectDuplicates to master

kuiktzecheng committed:
Add files via upload

kuiktzecheng committed:
Update README.md

kuiktzecheng committed:
Update README.md

kuiktzecheng committed:
Update README.md

kuiktzecheng committed:
Update README.md

kuiktzecheng committed:
Update README.md

kuiktzecheng committed:
Update README.md
